src/sas/sascalc/corfunc/transform_thread.py

```diff
@@ -12,32 +13,75 @@ def __init__(self, raw_data, extrapolated_data, bg, updatefn=None,
         self.background = bg
         self.extrapolation = extrapolated_data
 
+    def check_if_cancelled(self):
+        if self.isquit():
+            self.update("Fourier transform cancelled.")
+            self.complete(transforms=None)
+            return True
+        return False
+
     def compute(self):
         qs = self.extrapolation.x
         iqs = self.extrapolation.y
         q = self.data.x
         background = self.background
 
+        xs = np.pi*np.arange(len(qs),dtype=np.float32)/(q[1]-q[0])/len(qs)
+
         self.ready(delay=0.0)
-        self.update(msg="Starting Fourier transform.")
+        self.update(msg="Fourier transform in progress.")
         self.ready(delay=0.0)
-        if self.isquit():
-            return
+
+        if self.check_if_cancelled(): return
         try:
-            gamma = dct((iqs-background)*qs**2)
-            gamma = gamma / gamma.max()
-        except:
+            # ----- 1D Correlation Function -----
+            gamma1 = dct((iqs-background)*qs**2)
+            Q = gamma1.max()
+            gamma1 /= Q
+
+            if self.check_if_cancelled(): return
+
+            # ----- 3D Correlation Function -----
+            # gamma3(R) = 1/R int_{0}^{R} gamma1(x) dx
+            # trapz uses the trapezium rule to calculate the integral
+            mask = xs <= 200.0 # Only calculate gamma3 up to x=200 (as this is all that's plotted)
+            gamma3 = [trapz(gamma1[:n], xs[:n])/xs[n-1] for n in range(2, len(xs[mask]) + 1)]
```

Comment:
scipy.integrate.cumtrapz will be O(n) rather than O(n^2).  It also avoids looping in python instead of C so it would be much faster even if it were only O(n).